### DISCUSSION TITLE
Display does not work with device Sonoff POWR316D and Software tasmota32-bluetooth.bin

### DISCUSSION THREAD

**Discussion by mawinckie** [arendst/Tasmota#16933] [Q&A] [answered]
### PROBLEM DESCRIPTION
Display does not work with device Sonoff POWR316D and Software tasmota32-bluetooth.bin

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x ] Device used (e.g., Sonoff Basic): POWR316D
- [x] Tasmota binary firmware version number used:12.2.0.2 tasmota32-bluetooth.bin
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: ESP-Flasher 
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by mawinckie** [comment]
With software tasmota32.bin it work but with software tasmota32-bluetooth.bin display does not update anymore and show either nothing or if tasmota32.bin was flashed before the last sceen with software tasmota32.bin.

**Reply by sfromis** [reply, marked as answer]
That's correct, not all features are included in all precompiled binaries. This is not a "problem".

You can create a custom binary with the feature mix you want.

**Comment by mawinckie** [comment]
Now it works :-).
Good advice, thank you very much.

**Comment by crazyasyou** [comment]
I have the same display issue on my Sonoff POWR316D after upgrading from tasmota32.bin to tasmota32-bluetooth.bin (doubling up as a BLE MQTT proxy) - Can someone advise how to compile a custom .bin that would include what is required for bluetooth and a working disaplay (i've never complied tasmota, but I'm a quicker leaner)

**Reply by barbudor** [reply]
You need to create a variant which is based on tasmota32 and includes everything needed to add BLE

Add a file `platformio_tasmota_cenv.ini` in which you declare you new variant such as `tasmota32-pow_ble`:

```
[platformio]
default_envs =
    tasmota32-pow_ble

[env:tasmota32-tasmota32-pow_ble]
extends                 = env:tasmota32_base
build_flags             = ${env:tasmota32_base.build_flags}
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_ssl
```

Then in your `user_config_override.h` you need to add the features for BLE
At minimum:
```
    #define USE_BLE_ESP32                            // Enable new BLE driver
    #define BLE_ESP32_ENABLE                  true   // [SetOption115] Default value for SetOption115
    #define USE_MI_ESP32                             // (ESP32 only) Add support for ESP32 as a BLE-bridge (+9k2 mem, +292k flash)
```
You can also preconfigure some settings to make htem default and permanent.
These are mine but feel free to customize them
```
    #define MI32_PAGE                         10
    #define MI32_OPTION0_ALWAYS_AGGREGATE     true
    #define MI32_OPTION1_NO_SUMMARY           false
    #define MI32_OPTION2_DIRECT_BRIDGED_MODE  false
    #define MI32_OPTION4_IGNORE_BOGUS_BATTERY true
    #define MI32_OPTION5_ONLY_ALIASED         true
    #define MI32_OPTION6_MQTT_TYPE            2
    #define MI32_BLE_TOPIC "blegateway"
```

**Reply by crazyasyou** [reply]
Thanks for the reply, I'd already worked out the user_config_override.h section but didn't know about platformio_tasmota_cenv.ini - I've done what you've said (I'm using Gitpod Cloud) and when I try to compile using command `platformio run -e tasmota32-pow_ble` I just get an error _Error: Unknown environment names 'tasmota32-pow_ble'. Valid names are 'tasmota32-display_  the list is long but no sign of my custom pow-ble, any idea where I might have gone wrong?

![image](https://user-images.githubusercontent.com/6671537/225871728-21525384-d970-40d9-a0bf-67cc83908e49.png)

![image](https://user-images.githubusercontent.com/6671537/225871847-b1dbf1b5-02ea-4de2-b651-d2996778a069.png)

**Reply by sfromis** [reply]
It does not work to create an environment `tasmota32-tasmota32-pow_ble` and then try to build `tasmota32-pow_ble` instead.

**Reply by crazyasyou** [reply]
Thanks, I didn't spot the prefix 'tasmota32-' typo - That's compiled and successfully flashed on to my POW16D but the LCD display is still not rotating to show voltage/watts etc - Do I need to include more features ?

**Reply by barbudor** [reply]
Oups my bad, it's slighly different on Tasmota32 and Tasmota

You also need to add everything that you need from standard FIRMWARE_TASMOTA32:
This is what is inside FIRMWARE_TASMOTA32 : https://github.com/arendst/Tasmota/blob/d0f463761105ae11439bbdce1105f54494cbc775/tasmota/include/tasmota_configurations_ESP32.h#L546-L761
In case of doubt copy everything that is between `#ifdef FIRMWARE_TASMOTA32` and the last `#endif`

Actually you may want to remove things that you would obvioulsy don't need as there are actually a LOT of things in Tamsota32

**Comment by JFHCFC** [comment]
Display (TM1621) works on 12.5.0(tasmota32), higher versions is to big

http://ota.tasmota.com/tasmota32/release-12.5.0/tasmota32.bin

`{"NAME":"Sonoff POWR316D","GPIO":[32,0,0,0,0,576,0,0,0,224,9280,0,3104,0,320,0,0,0,0,0,0,9184,9248,9216,0,0,0,0,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}`

https://templates.blakadder.com/sonoff_POWR316D.html

**Reply by sfromis** [reply]
No, newer versions are **not** "too big", but you need to upgrade to the safeboot partition scheme. https://github.com/arendst/Tasmota/discussions/18983
